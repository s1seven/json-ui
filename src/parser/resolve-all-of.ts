import omit from "lodash/omit";
import mapValues from "lodash/mapValues";
import isArray from "lodash/isArray";
import isObject from "lodash/isObject";
import type { ExtractObjects, JSONSchema7Value } from "../utils/helper-types";
import { deepMerge } from "../utils/deep-merge";

export const resolveAllOf = <T extends JSONSchema7Value>(item: T): T =>
  !isObject(item)
    ? item
    : isArray(item)
    ? (item.map(resolveAllOf) as T)
    : (mapValues(
        isArray(item.allOf)
          ? deepMerge<ExtractObjects<JSONSchema7Value>>(
              ...(item.allOf.map(
                resolveAllOf
              ) as ExtractObjects<JSONSchema7Value>[]),
              omit(item, "allOf")
            )
          : item,
        resolveAllOf
      ) as T);
