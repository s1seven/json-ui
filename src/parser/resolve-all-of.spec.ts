import { JSONSchema7 } from "json-schema";
import { resolveAllOf } from "./resolve-all-of";

test("it matches the snapshot", async () => {
  const schema = (await import("../../fixtures/schema-with-all-of.json"))
    .default as unknown as JSONSchema7;
  const result = resolveAllOf(schema);
  expect(result).toMatchSnapshot();
});
