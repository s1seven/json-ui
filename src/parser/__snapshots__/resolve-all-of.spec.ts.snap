// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot 1`] = `
{
  "$id": "https://example.com/schemas/product",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "properties": {
    "category": {
      "enum": [
        "book",
        "electronics",
        "clothing",
      ],
      "type": "string",
    },
    "name": {
      "type": "string",
    },
    "price": {
      "minimum": 0,
      "type": "number",
    },
    "rating": {
      "maximum": 5,
      "minimum": 1,
      "type": "number",
    },
  },
  "required": [
    "name",
    "price",
    "category",
  ],
  "type": "object",
}
`;
